## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#0053c394-8eb1-4be2-a18a-8c9ef4e4ef9d) |
| mint-erasure.sh | :x: [more...](#0ffdd2ee-c905-4233-bb2d-b34498ba11ad) |
| mint-dist-erasure.sh | :x: [more...](#61324280-6628-40f5-b9b4-c5f5d32dc769) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#760ca610-6e2d-48aa-ae2c-c31e4d4fbdcd) |
<a name="0053c394-8eb1-4be2-a18a-8c9ef4e4ef9d"></a>
### 13544-20bb4e0/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s6:31825
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/16) Running aws-sdk-go tests ... done in 1 seconds
(2/16) Running aws-sdk-java tests ... done in 1 seconds
(3/16) Running aws-sdk-php tests ... done in 44 seconds
(4/16) Running aws-sdk-ruby tests ... done in 3 seconds
(5/16) Running awscli tests ... done in 1 minutes and 0 seconds
(6/16) Running healthcheck tests ... done in 0 seconds
(7/16) Running mc tests ... done in 25 seconds
(8/16) Running minio-dotnet tests ... done in 51 seconds
(9/16) Running minio-go tests ... done in 1 minutes and 17 seconds
(10/16) Running minio-java tests ... FAILED in 5 minutes and 29 seconds
[java.io.IOException: unexpected end of stream on http://minio-k8s6:31825/..., java.io.IOException: unexpected end of stream on http://minio-k8s6:31825/..., java.io.IOException: unexpected end of stream on http://minio-k8s6:31825/...]
(10/16) Running minio-js tests ... FAILED in 35 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 6,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:665:10) at endReadableNT (internal/streams/readable.js:1317:12) at processTicksAndRejections (internal/process/task_queues.js:82:21)"
}
(10/16) Running minio-py tests ... FAILED in 2 minutes and 2 seconds
{
  "name": "minio-py:test_remove_objects",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-56656db7-5fde-44a1-8dfc-e88a2ce68da7",
    "delete_object_list": [
      "prefix-0",
      "prefix-1",
      "prefix-2",
      "prefix-3",
      "prefix-4",
      "prefix-5",
      "prefix-6",
      "prefix-7",
      "prefix-8",
      "prefix-9"
    ]
  },
  "message": "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 706, in urlopen\n    chunked=chunked,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 445, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 440, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.6/http/client.py\", line 280, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nhttp.client.RemoteDisconnected: Remote end closed connection without response\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1862, in test_remove_objects\n    _test_remove_objects(log_entry)\n  File \"/mint/run/core/minio-py/tests.py\", line 1855, in _test_remove_objects\n    for err in errs:\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2043, in remove_objects\n    bypass_governance_mode=bypass_governance_mode,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1977, in _delete_objects\n    query_params={\"delete\": \"\"},\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 422, in _execute\n    no_body_trace=no_body_trace,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 285, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 756, in urlopen\n    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 532, in increment\n    raise six.reraise(type(error), error, _stacktrace)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/packages/six.py\", line 769, in reraise\n    raise value.with_traceback(tb)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 706, in urlopen\n    chunked=chunked,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 445, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 440, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.6/http/client.py\", line 280, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nurllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))\n",
  "duration": 260
}
(10/16) Running s3cmd tests ... done in 3 minutes and 39 seconds
(11/16) Running s3select tests ... done in 13 seconds
(12/16) Running security tests ... done in 0 seconds
(13/16) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 13 out of 16 tests successfully.
```
<a name="0ffdd2ee-c905-4233-bb2d-b34498ba11ad"></a>
### 13544-20bb4e0/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s4:32734
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/16) Running aws-sdk-go tests ... done in 0 seconds
(2/16) Running aws-sdk-java tests ... done in 1 seconds
(3/16) Running aws-sdk-php tests ... done in 43 seconds
(4/16) Running aws-sdk-ruby tests ... done in 2 seconds
(5/16) Running awscli tests ... done in 1 minutes and 58 seconds
(6/16) Running healthcheck tests ... done in 0 seconds
(7/16) Running mc tests ... done in 19 seconds
(8/16) Running minio-dotnet tests ... done in 46 seconds
(9/16) Running minio-go tests ... done in 53 seconds
(10/16) Running minio-java tests ... FAILED in 14 seconds
[java.io.IOException: unexpected end of stream on http://minio-k8s4:32734/..., java.io.IOException: unexpected end of stream on http://minio-k8s4:32734/..., java.io.IOException: unexpected end of stream on http://minio-k8s4:32734/..., java.io.IOException: unexpected end of stream on http://minio-k8s4:32734/..., java.io.IOException: unexpected end of stream on http://minio-k8s4:32734/...]
(10/16) Running minio-js tests ... FAILED in 1 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 2,
  "status": "FAIL",
  "error": "Error: connect ECONNREFUSED 15.15.15.4:32734 at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1148:16)"
}
(10/16) Running minio-py tests ... FAILED in 3 minutes and 0 seconds
{
  "name": "minio-py:test_remove_objects",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-1d433d27-bb34-43ff-b46c-f432355293d4",
    "delete_object_list": [
      "prefix-0",
      "prefix-1",
      "prefix-2",
      "prefix-3",
      "prefix-4",
      "prefix-5",
      "prefix-6",
      "prefix-7",
      "prefix-8",
      "prefix-9"
    ]
  },
  "message": "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 706, in urlopen\n    chunked=chunked,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 445, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 440, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.6/http/client.py\", line 280, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nhttp.client.RemoteDisconnected: Remote end closed connection without response\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1862, in test_remove_objects\n    _test_remove_objects(log_entry)\n  File \"/mint/run/core/minio-py/tests.py\", line 1855, in _test_remove_objects\n    for err in errs:\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2043, in remove_objects\n    bypass_governance_mode=bypass_governance_mode,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1977, in _delete_objects\n    query_params={\"delete\": \"\"},\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 422, in _execute\n    no_body_trace=no_body_trace,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 285, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 756, in urlopen\n    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 532, in increment\n    raise six.reraise(type(error), error, _stacktrace)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/packages/six.py\", line 769, in reraise\n    raise value.with_traceback(tb)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 706, in urlopen\n    chunked=chunked,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 445, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 440, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.6/http/client.py\", line 280, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nurllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))\n",
  "duration": 191
}
(10/16) Running s3cmd tests ... FAILED in 0 seconds
{
  "name": "s3cmd",
  "duration": "248",
  "function": "test_make_bucket",
  "status": "FAIL",
  "error": "\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n    An unexpected error has occurred.\n  Please try reproducing the error using\n  the latest s3cmd code from the git master\n  branch found at:\n    https://github.com/s3tools/s3cmd\n  and have a look at the known issues list:\n    https://github.com/s3tools/s3cmd/wiki/Common-known-issues-and-their-solutions\n  If the error persists, please report the\n  following lines (removing any private\n  info as necessary) to:\n   s3tools-bugs@lists.sourceforge.net\n\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n\nInvoked as: /usr/local/bin/s3cmd --config /tmp/.s3cmd-28291/s3cfg mb s3://s3cmd-test-bucket-17421\nProblem: <class 'ConnectionRefusedError: [Errno 111] Connection refused\nS3cmd:   2.1.0\npython:   3.6.9 (default, Jan 26 2021, 15:33:00) \n[GCC 8.4.0]\nenvironment LANG=C.UTF-8\n\nTraceback (most recent call last):\n  File \"/usr/local/bin/s3cmd\", line 3121, in <module>\n    rc = main()\n  File \"/usr/local/bin/s3cmd\", line 3030, in main\n    rc = cmd_func(args)\n  File \"/usr/local/bin/s3cmd\", line 251, in cmd_bucket_create\n    response = s3.bucket_create(uri.bucket(), cfg.bucket_location)\n  File \"/usr/local/lib/python3.6/dist-packages/S3/S3.py\", line 408, in bucket_create\n    response = self.send_request(request)\n  File \"/usr/local/lib/python3.6/dist-packages/S3/S3.py\", line 1281, in send_request\n    conn = ConnMan.get(self.get_hostname(resource['bucket']))\n  File \"/usr/local/lib/python3.6/dist-packages/S3/ConnMan.py\", line 251, in get\n    conn.c.connect()\n  File \"/usr/lib/python3.6/http/client.py\", line 952, in connect\n    (self.host,self.port), self.timeout, self.source_address)\n  File \"/usr/lib/python3.6/socket.py\", line 724, in create_connection\n    raise err\n  File \"/usr/lib/python3.6/socket.py\", line 713, in create_connection\n    sock.connect(sa)\nConnectionRefusedError: [Errno 111] Connection refused\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n    An unexpected error has occurred.\n  Please try reproducing the error using\n  the latest s3cmd code from the git master\n  branch found at:\n    https://github.com/s3tools/s3cmd\n  and have a look at the known issues list:\n    https://github.com/s3tools/s3cmd/wiki/Common-known-issues-and-their-solutions\n  If the error persists, please report the\n  above lines (removing any private\n  info as necessary) to:\n   s3tools-bugs@lists.sourceforge.net\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!"
}
(10/16) Running s3select tests ... FAILED in 7 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, request)",
  "args": {
    "bucket_name": "s3select-test-35539943-2efa-48ef-a17f-8a5299871891"
  },
  "duration": 7041,
  "message": "HTTPConnectionPool(host='minio-k8s4', port=32734): Max retries exceeded with url: /s3select-test-35539943-2efa-48ef-a17f-8a5299871891 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8d7d1271d0>: Failed to establish a new connection: [Errno 111] Connection refused',))",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connection.py\", line 170, in _new_conn\n    (self._dns_host, self.port), self.timeout, **extra_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/connection.py\", line 96, in create_connection\n    raise err\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/connection.py\", line 86, in create_connection\n    sock.connect(sa)\nConnectionRefusedError: [Errno 111] Connection refused\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 706, in urlopen\n    chunked=chunked,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 394, in _make_request\n    conn.request(method, url, **httplib_request_kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connection.py\", line 234, in request\n    super(HTTPConnection, self).request(method, url, body=body, headers=headers)\n  File \"/usr/lib/python3.6/http/client.py\", line 1281, in request\n    self._send_request(method, url, body, headers, encode_chunked)\n  File \"/usr/lib/python3.6/http/client.py\", line 1327, in _send_request\n    self.endheaders(body, encode_chunked=encode_chunked)\n  File \"/usr/lib/python3.6/http/client.py\", line 1276, in endheaders\n    self._send_output(message_body, encode_chunked=encode_chunked)\n  File \"/usr/lib/python3.6/http/client.py\", line 1042, in _send_output\n    self.send(msg)\n  File \"/usr/lib/python3.6/http/client.py\", line 980, in send\n    self.connect()\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connection.py\", line 200, in connect\n    conn = self._new_conn()\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connection.py\", line 182, in _new_conn\n    self, \"Failed to establish a new connection: %s\" % e\nurllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f8d7d1271d0>: Failed to establish a new connection: [Errno 111] Connection refused\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 53, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 95, in test_csv_input_custom_quote_char\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 638, in make_bucket\n    headers=headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 285, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 796, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 796, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 796, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 756, in urlopen\n    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 574, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-k8s4', port=32734): Max retries exceeded with url: /s3select-test-35539943-2efa-48ef-a17f-8a5299871891 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8d7d1271d0>: Failed to establish a new connection: [Errno 111] Connection refused',))\n",
  "status": "FAIL"
}
(10/16) Running security tests ... done in 0 seconds
(11/16) Running versioning tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "versioning-test-kn1nbkxt21zlq0"
  },
  "duration": 361,
  "error": "RequestError: send request failed\ncaused by: Put \"http://minio-k8s4:32734/versioning-test-kn1nbkxt21zlq0\": dial tcp 15.15.15.4:32734: connect: connection refused",
  "function": "/mint/build/versioning/bucket.go:47: main.testMakeBucket",
  "message": "Versioning CreateBucket Failed",
  "name": "versioning",
  "status": "FAIL"
}

Executed 10 out of 16 tests successfully.
```
<a name="61324280-6628-40f5-b9b4-c5f5d32dc769"></a>
### 13544-20bb4e0/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s4:32177
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/16) Running aws-sdk-go tests ... done in 1 seconds
(2/16) Running aws-sdk-java tests ... done in 1 seconds
(3/16) Running aws-sdk-php tests ... done in 43 seconds
(4/16) Running aws-sdk-ruby tests ... done in 3 seconds
(5/16) Running awscli tests ... done in 1 minutes and 55 seconds
(6/16) Running healthcheck tests ... done in 1 seconds
(7/16) Running mc tests ... done in 24 seconds
(8/16) Running minio-dotnet tests ... done in 49 seconds
(9/16) Running minio-go tests ... done in 1 minutes and 16 seconds
(10/16) Running minio-java tests ... FAILED in 4 minutes and 4 seconds
[java.io.IOException: unexpected end of stream on http://minio-k8s4:32177/..., java.io.IOException: unexpected end of stream on http://minio-k8s4:32177/..., java.io.IOException: unexpected end of stream on http://minio-k8s4:32177/..., java.io.IOException: unexpected end of stream on http://minio-k8s4:32177/..., java.io.IOException: unexpected end of stream on http://minio-k8s4:32177/...]
(10/16) Running minio-js tests ... FAILED in 21 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 10,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:665:10) at endReadableNT (internal/streams/readable.js:1317:12) at processTicksAndRejections (internal/process/task_queues.js:82:21)"
}
(10/16) Running minio-py tests ... FAILED in 3 minutes and 24 seconds
{
  "name": "minio-py:test_remove_objects",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-2a4ebdb8-6a8c-4817-b8a0-c4e1af1ed4eb",
    "delete_object_list": [
      "prefix-0",
      "prefix-1",
      "prefix-2",
      "prefix-3",
      "prefix-4",
      "prefix-5",
      "prefix-6",
      "prefix-7",
      "prefix-8",
      "prefix-9"
    ]
  },
  "message": "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 706, in urlopen\n    chunked=chunked,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 445, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 440, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.6/http/client.py\", line 280, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nhttp.client.RemoteDisconnected: Remote end closed connection without response\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1862, in test_remove_objects\n    _test_remove_objects(log_entry)\n  File \"/mint/run/core/minio-py/tests.py\", line 1855, in _test_remove_objects\n    for err in errs:\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2043, in remove_objects\n    bypass_governance_mode=bypass_governance_mode,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1977, in _delete_objects\n    query_params={\"delete\": \"\"},\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 422, in _execute\n    no_body_trace=no_body_trace,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 285, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 756, in urlopen\n    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 532, in increment\n    raise six.reraise(type(error), error, _stacktrace)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/packages/six.py\", line 769, in reraise\n    raise value.with_traceback(tb)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 706, in urlopen\n    chunked=chunked,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 445, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 440, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.6/http/client.py\", line 280, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nurllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))\n",
  "duration": 2235
}
(10/16) Running s3cmd tests ... FAILED in 32 seconds
{
  "name": "s3cmd",
  "duration": "14313",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "WARNING: Bucket is not empty. Removing all the objects from it first. This may take some time...\ndelete: 's3://s3cmd-test-bucket-19776/datafile-0-b'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-1-MB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-1-b'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-1-kB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-1.03-MB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-10-MB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-10-kB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-100-kB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-11-MB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-129-MB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-33-kB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-5-MB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-5243880-b'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-6-MB'\ndelete: 's3://s3cmd-test-bucket-19776/datafile-65-MB'\nERROR: S3 error: 409 (BucketNotEmpty): The bucket you tried to delete is not empty"
}
(10/16) Running s3select tests ... done in 2 minutes and 25 seconds
(11/16) Running security tests ... done in 0 seconds
(12/16) Running versioning tests ... done in 4 minutes and 4 seconds

Executed 12 out of 16 tests successfully.
```
<a name="760ca610-6e2d-48aa-ae2c-c31e4d4fbdcd"></a>
### 13544-20bb4e0/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s6:32005
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/16) Running aws-sdk-go tests ... done in 1 seconds
(2/16) Running aws-sdk-java tests ... done in 1 seconds
(3/16) Running aws-sdk-php tests ... done in 43 seconds
(4/16) Running aws-sdk-ruby tests ... done in 4 seconds
(5/16) Running awscli tests ... done in 2 minutes and 1 seconds
(6/16) Running healthcheck tests ... done in 0 seconds
(7/16) Running mc tests ... done in 24 seconds
(8/16) Running minio-dotnet tests ... done in 50 seconds
(9/16) Running minio-go tests ... done in 1 minutes and 32 seconds
(10/16) Running minio-java tests ... FAILED in 2 minutes and 8 seconds
[java.io.IOException: unexpected end of stream on http://minio-k8s6:32005/..., java.io.IOException: unexpected end of stream on http://minio-k8s6:32005/..., java.io.IOException: unexpected end of stream on http://minio-k8s6:32005/..., java.io.IOException: unexpected end of stream on http://minio-k8s6:32005/..., java.io.IOException: unexpected end of stream on http://minio-k8s6:32005/...]
(10/16) Running minio-js tests ... done in 52 seconds
(11/16) Running minio-py tests ... FAILED in 1 minutes and 27 seconds
{
  "name": "minio-py:test_remove_objects",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-b74bc21d-fbdc-472c-873b-26e4c6a6c641",
    "delete_object_list": [
      "prefix-0",
      "prefix-1",
      "prefix-2",
      "prefix-3",
      "prefix-4",
      "prefix-5",
      "prefix-6",
      "prefix-7",
      "prefix-8",
      "prefix-9"
    ]
  },
  "message": "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 706, in urlopen\n    chunked=chunked,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 445, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 440, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.6/http/client.py\", line 280, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nhttp.client.RemoteDisconnected: Remote end closed connection without response\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1862, in test_remove_objects\n    _test_remove_objects(log_entry)\n  File \"/mint/run/core/minio-py/tests.py\", line 1855, in _test_remove_objects\n    for err in errs:\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2043, in remove_objects\n    bypass_governance_mode=bypass_governance_mode,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1977, in _delete_objects\n    query_params={\"delete\": \"\"},\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 422, in _execute\n    no_body_trace=no_body_trace,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 285, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 756, in urlopen\n    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 532, in increment\n    raise six.reraise(type(error), error, _stacktrace)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/packages/six.py\", line 769, in reraise\n    raise value.with_traceback(tb)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 706, in urlopen\n    chunked=chunked,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 445, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 440, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.6/http/client.py\", line 280, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nurllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))\n",
  "duration": 252
}
(11/16) Running s3cmd tests ... done in 1 minutes and 49 seconds
(12/16) Running s3select tests ... done in 36 seconds
(13/16) Running security tests ... done in 0 seconds
(14/16) Running versioning tests ... done in 4 minutes and 4 seconds

Executed 14 out of 16 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/erasure-object.go
+++ b/cmd/erasure-object.go
@@ -1088,37 +1088,64 @@ func (er erasureObjects) DeleteObjects(ctx context.Context, bucket string, objec
 		writeQuorums[i] = getWriteQuorum(len(storageDisks))
 	}
 
-	versions := make([]FileInfo, len(objects))
+	versionsMap := make(map[string]FileInfoVersions, len(objects))
 	for i := range objects {
-		if objects[i].VersionID == "" {
-			modTime := opts.MTime
-			if opts.MTime.IsZero() {
-				modTime = UTCNow()
-			}
-			uuid := opts.VersionID
-			if uuid == "" {
-				uuid = mustGetUUID()
-			}
-			if opts.Versioned || opts.VersionSuspended {
-				versions[i] = FileInfo{
-					Name:             objects[i].ObjectName,
-					ModTime:          modTime,
-					Deleted:          true, // delete marker
-					ReplicationState: objects[i].ReplicationState(),
-				}
-				versions[i].SetTierFreeVersionID(mustGetUUID())
-				if opts.Versioned {
-					versions[i].VersionID = uuid
-				}
-				continue
-			}
-		}
-		versions[i] = FileInfo{
+		// Construct the FileInfo data that needs to be preserved on the disk.
+		vr := FileInfo{
 			Name:             objects[i].ObjectName,
 			VersionID:        objects[i].VersionID,
 			ReplicationState: objects[i].ReplicationState(),
+			// save the index to set correct error at this index.
+			Idx: i,
 		}
-		versions[i].SetTierFreeVersionID(mustGetUUID())
+		vr.SetTierFreeVersionID(mustGetUUID())
+		// VersionID is not set means delete is not specific about
+		// any version, look for if the bucket is versioned or not.
+		if objects[i].VersionID == "" {
+			if opts.Versioned || opts.VersionSuspended {
+				// Bucket is versioned and no version was explicitly
+				// mentioned for deletes, create a delete marker instead.
+				vr.ModTime = UTCNow()
+				vr.Deleted = true
+				// Versioning suspended means that we add a `null` version
+				// delete marker, if not add a new version for this delete
+				// marker.
+				if opts.Versioned {
+					vr.VersionID = mustGetUUID()
+				}
+			}
+		}
+		// De-dup same object name to collect multiple versions for same object.
+		v, ok := versionsMap[objects[i].ObjectName]
+		if ok {
+			v.Versions = append(v.Versions, vr)
+		} else {
+			v = FileInfoVersions{
+				Name:     vr.Name,
+				Versions: []FileInfo{vr},
+			}
+		}
+		if vr.Deleted {
+			dobjects[i] = DeletedObject{
+				DeleteMarker:          vr.Deleted,
+				DeleteMarkerVersionID: vr.VersionID,
+				DeleteMarkerMTime:     DeleteMarkerMTime{vr.ModTime},
+				ObjectName:            vr.Name,
+				ReplicationState:      vr.ReplicationState,
+			}
+		} else {
+			dobjects[i] = DeletedObject{
+				ObjectName:       vr.Name,
+				VersionID:        vr.VersionID,
+				ReplicationState: vr.ReplicationState,
+			}
+		}
+		versionsMap[objects[i].ObjectName] = v
+	}
+
+	dedupVersions := make([]FileInfoVersions, 0, len(versionsMap))
+	for _, version := range versionsMap {
+		dedupVersions = append(dedupVersions, version)
 	}
 
 	// Initialize list of errors.
@@ -1130,17 +1157,24 @@ func (er erasureObjects) DeleteObjects(ctx context.Context, bucket string, objec
 		wg.Add(1)
 		go func(index int, disk StorageAPI) {
 			defer wg.Done()
+			delObjErrs[index] = make([]error, len(objects))
 			if disk == nil {
-				delObjErrs[index] = make([]error, len(versions))
-				for i := range versions {
+				for i := range objects {
 					delObjErrs[index][i] = errDiskNotFound
 				}
 				return
 			}
-			delObjErrs[index] = disk.DeleteVersions(ctx, bucket, versions)
+			errs := disk.DeleteVersions(ctx, bucket, dedupVersions)
+			for i, err := range errs {
+				if err == nil {
+					continue
+				}
+				for _, v := range dedupVersions[i].Versions {
+					delObjErrs[index][v.Idx] = err
+				}
+			}
 		}(index, disk)
 	}
-
 	wg.Wait()
 
 	// Reduce errors for each object
@@ -1162,28 +1196,17 @@ func (er erasureObjects) DeleteObjects(ctx context.Context, bucket string, objec
 		}
 
 		if errs[objIndex] == nil {
-			NSUpdated(bucket, objects[objIndex].ObjectName)
-		}
-
-		if versions[objIndex].Deleted {
-			dobjects[objIndex] = DeletedObject{
-				DeleteMarker:          versions[objIndex].Deleted,
-				DeleteMarkerVersionID: versions[objIndex].VersionID,
-				DeleteMarkerMTime:     DeleteMarkerMTime{versions[objIndex].ModTime},
-				ObjectName:            versions[objIndex].Name,
-				ReplicationState:      versions[objIndex].ReplicationState,
-			}
-		} else {
-			dobjects[objIndex] = DeletedObject{
-				ObjectName:       versions[objIndex].Name,
-				VersionID:        versions[objIndex].VersionID,
-				ReplicationState: versions[objIndex].ReplicationState,
-			}
+			defer NSUpdated(bucket, objects[objIndex].ObjectName)
 		}
 	}
 
 	// Check failed deletes across multiple objects
-	for _, version := range versions {
+	for i, dobj := range dobjects {
+		// This object errored, no need to attempt a heal.
+		if errs[i] != nil {
+			continue
+		}
+
 		// Check if there is any offline disk and add it to the MRF list
 		for _, disk := range storageDisks {
 			if disk != nil && disk.IsOnline() {
@@ -1193,7 +1216,7 @@ func (er erasureObjects) DeleteObjects(ctx context.Context, bucket string, objec
 
 			// all other direct versionId references we should
 			// ensure no dangling file is left over.
-			er.addPartial(bucket, version.Name, version.VersionID, -1)
+			er.addPartial(bucket, dobj.ObjectName, dobj.VersionID, -1)
 			break
 		}
 	}

--- a/cmd/naughty-disk_test.go
+++ b/cmd/naughty-disk_test.go
@@ -225,7 +225,7 @@ func (d *naughtyDisk) Delete(ctx context.Context, volume string, path string, re
 	return d.disk.Delete(ctx, volume, path, recursive)
 }
 
-func (d *naughtyDisk) DeleteVersions(ctx context.Context, volume string, versions []FileInfo) []error {
+func (d *naughtyDisk) DeleteVersions(ctx context.Context, volume string, versions []FileInfoVersions) []error {
 	if err := d.calcError(); err != nil {
 		errs := make([]error, len(versions))
 		for i := range errs {

--- a/cmd/storage-datatypes.go
+++ b/cmd/storage-datatypes.go
@@ -80,20 +80,20 @@ type FilesInfoVersions struct {
 }
 
 // FileInfoVersions represent a list of versions for a given file.
+//msgp:tuple FileInfoVersions
+// The above means that any added/deleted fields are incompatible.
 type FileInfoVersions struct {
 	// Name of the volume.
-	Volume string
+	Volume string `msg:"v,omitempty"`
 
 	// Name of the file.
-	Name string
-
-	IsEmptyDir bool
+	Name string `msg:"n,omitempty"`
 
 	// Represents the latest mod time of the
 	// latest version.
-	LatestModTime time.Time
+	LatestModTime time.Time `msg:"lm"`
 
-	Versions []FileInfo
+	Versions []FileInfo `msg:"vs"`
 }
 
 // findVersionIndex will return the version index where the version
@@ -115,69 +115,74 @@ func (f *FileInfoVersions) findVersionIndex(v string) int {
 // The above means that any added/deleted fields are incompatible.
 type FileInfo struct {
 	// Name of the volume.
-	Volume string
+	Volume string `msg:"v,omitempty"`
 
 	// Name of the file.
-	Name string
+	Name string `msg:"n,omitempty"`
 
 	// Version of the file.
-	VersionID string
+	VersionID string `msg:"vid,omitempty"`
 
 	// Indicates if the version is the latest
-	IsLatest bool
+	IsLatest bool `msg:"is"`
 
 	// Deleted is set when this FileInfo represents
 	// a deleted marker for a versioned bucket.
-	Deleted bool
+	Deleted bool `msg:"del"`
 
 	// TransitionStatus is set to Pending/Complete for transitioned
 	// entries based on state of transition
-	TransitionStatus string
+	TransitionStatus string `msg:"ts"`
 	// TransitionedObjName is the object name on the remote tier corresponding
 	// to object (version) on the source tier.
-	TransitionedObjName string
+	TransitionedObjName string `msg:"to"`
 	// TransitionTier is the storage class label assigned to remote tier.
-	TransitionTier string
+	TransitionTier string `msg:"tt"`
 	// TransitionVersionID stores a version ID of the object associate
 	// with the remote tier.
-	TransitionVersionID string
+	TransitionVersionID string `msg:"tv"`
 	// ExpireRestored indicates that the restored object is to be expired.
-	ExpireRestored bool
+	ExpireRestored bool `msg:"exp"`
 
 	// DataDir of the file
-	DataDir string
+	DataDir string `msg:"dd"`
 
 	// Indicates if this object is still in V1 format.
-	XLV1 bool
+	XLV1 bool `msg:"v1"`
 
 	// Date and time when the file was last modified, if Deleted
 	// is 'true' this value represents when while was deleted.
-	ModTime time.Time
+	ModTime time.Time `msg:"mt"`
 
 	// Total file size.
-	Size int64
+	Size int64 `msg:"sz"`
 
 	// File mode bits.
-	Mode uint32
+	Mode uint32 `msg:"m"`
 
 	// File metadata
-	Metadata map[string]string
+	Metadata map[string]string `msg:"meta"`
 
 	// All the parts per object.
-	Parts []ObjectPartInfo
+	Parts []ObjectPartInfo `msg:"parts"`
 
 	// Erasure info for all objects.
-	Erasure ErasureInfo
+	Erasure ErasureInfo `msg:"ei"`
 
-	MarkDeleted      bool             // mark this version as deleted
-	ReplicationState ReplicationState // Internal replication state to be passed back in ObjectInfo
+	MarkDeleted      bool             `msg:"md"` // mark this version as deleted
+	ReplicationState ReplicationState `msg:"rs"` // Internal replication state to be passed back in ObjectInfo
 
-	Data []byte // optionally carries object data
+	Data []byte `msg:"d,allownil"` // optionally carries object data
 
-	NumVersions      int
-	SuccessorModTime time.Time
+	NumVersions      int       `msg:"nv"`
+	SuccessorModTime time.Time `msg:"smt"`
 
-	Fresh bool // indicates this is a first time call to write FileInfo.
+	Fresh bool `msg:"fr"` // indicates this is a first time call to write FileInfo.
+
+	// Position of this version or object in a multi-object delete call,
+	// no other caller must set this value other than multi-object delete call.
+	// usage in other calls in undefined please avoid.
+	Idx int `msg:"i"`
 }
 
 // InlineData returns true if object contents are inlined alongside its metadata.

--- a/cmd/storage-datatypes_gen.go
+++ b/cmd/storage-datatypes_gen.go
@@ -550,8 +550,8 @@ func (z *FileInfo) DecodeMsg(dc *msgp.Reader) (err error) {
 		err = msgp.WrapError(err)
 		return
 	}
-	if zb0001 != 24 {
-		err = msgp.ArrayError{Wanted: 24, Got: zb0001}
+	if zb0001 != 25 {
+		err = msgp.ArrayError{Wanted: 25, Got: zb0001}
 		return
 	}
 	z.Volume, err = dc.ReadString()
@@ -711,13 +711,18 @@ func (z *FileInfo) DecodeMsg(dc *msgp.Reader) (err error) {
 		err = msgp.WrapError(err, "Fresh")
 		return
 	}
+	z.Idx, err = dc.ReadInt()
+	if err != nil {
+		err = msgp.WrapError(err, "Idx")
+		return
+	}
 	return
 }
 
 // EncodeMsg implements msgp.Encodable
 func (z *FileInfo) EncodeMsg(en *msgp.Writer) (err error) {
-	// array header, size 24
-	err = en.Append(0xdc, 0x0, 0x18)
+	// array header, size 25
+	err = en.Append(0xdc, 0x0, 0x19)
 	if err != nil {
 		return
 	}
@@ -860,14 +865,19 @@ func (z *FileInfo) EncodeMsg(en *msgp.Writer) (err error) {
 		err = msgp.WrapError(err, "Fresh")
 		return
 	}
+	err = en.WriteInt(z.Idx)
+	if err != nil {
+		err = msgp.WrapError(err, "Idx")
+		return
+	}
 	return
 }
 
 // MarshalMsg implements msgp.Marshaler
 func (z *FileInfo) MarshalMsg(b []byte) (o []byte, err error) {
 	o = msgp.Require(b, z.Msgsize())
-	// array header, size 24
-	o = append(o, 0xdc, 0x0, 0x18)
+	// array header, size 25
+	o = append(o, 0xdc, 0x0, 0x19)
 	o = msgp.AppendString(o, z.Volume)
 	o = msgp.AppendString(o, z.Name)
 	o = msgp.AppendString(o, z.VersionID)
@@ -911,6 +921,7 @@ func (z *FileInfo) MarshalMsg(b []byte) (o []byte, err error) {
 	o = msgp.AppendInt(o, z.NumVersions)
 	o = msgp.AppendTime(o, z.SuccessorModTime)
 	o = msgp.AppendBool(o, z.Fresh)
+	o = msgp.AppendInt(o, z.Idx)
 	return
 }
 
@@ -922,8 +933,8 @@ func (z *FileInfo) UnmarshalMsg(bts []byte) (o []byte, err error) {
 		err = msgp.WrapError(err)
 		return
 	}
-	if zb0001 != 24 {
-		err = msgp.ArrayError{Wanted: 24, Got: zb0001}
+	if zb0001 != 25 {
+		err = msgp.ArrayError{Wanted: 25, Got: zb0001}
 		return
 	}
 	z.Volume, bts, err = msgp.ReadStringBytes(bts)
@@ -1083,6 +1094,11 @@ func (z *FileInfo) UnmarshalMsg(bts []byte) (o []byte, err error) {
 		err = msgp.WrapError(err, "Fresh")
 		return
 	}
+	z.Idx, bts, err = msgp.ReadIntBytes(bts)
+	if err != nil {
+		err = msgp.WrapError(err, "Idx")
+		return
+	}
 	o = bts
 	return
 }
@@ -1100,77 +1116,53 @@ func (z *FileInfo) Msgsize() (s int) {
 	for za0003 := range z.Parts {
 		s += z.Parts[za0003].Msgsize()
 	}
-	s += z.Erasure.Msgsize() + msgp.BoolSize + z.ReplicationState.Msgsize() + msgp.BytesPrefixSize + len(z.Data) + msgp.IntSize + msgp.TimeSize + msgp.BoolSize
+	s += z.Erasure.Msgsize() + msgp.BoolSize + z.ReplicationState.Msgsize() + msgp.BytesPrefixSize + len(z.Data) + msgp.IntSize + msgp.TimeSize + msgp.BoolSize + msgp.IntSize
 	return
 }
 
 // DecodeMsg implements msgp.Decodable
 func (z *FileInfoVersions) DecodeMsg(dc *msgp.Reader) (err error) {
-	var field []byte
-	_ = field
 	var zb0001 uint32
-	zb0001, err = dc.ReadMapHeader()
+	zb0001, err = dc.ReadArrayHeader()
 	if err != nil {
 		err = msgp.WrapError(err)
 		return
 	}
-	for zb0001 > 0 {
-		zb0001--
-		field, err = dc.ReadMapKeyPtr()
+	if zb0001 != 4 {
+		err = msgp.ArrayError{Wanted: 4, Got: zb0001}
+		return
+	}
+	z.Volume, err = dc.ReadString()
+	if err != nil {
+		err = msgp.WrapError(err, "Volume")
+		return
+	}
+	z.Name, err = dc.ReadString()
+	if err != nil {
+		err = msgp.WrapError(err, "Name")
+		return
+	}
+	z.LatestModTime, err = dc.ReadTime()
+	if err != nil {
+		err = msgp.WrapError(err, "LatestModTime")
+		return
+	}
+	var zb0002 uint32
+	zb0002, err = dc.ReadArrayHeader()
+	if err != nil {
+		err = msgp.WrapError(err, "Versions")
+		return
+	}
+	if cap(z.Versions) >= int(zb0002) {
+		z.Versions = (z.Versions)[:zb0002]
+	} else {
+		z.Versions = make([]FileInfo, zb0002)
+	}
+	for za0001 := range z.Versions {
+		err = z.Versions[za0001].DecodeMsg(dc)
 		if err != nil {
-			err = msgp.WrapError(err)
+			err = msgp.WrapError(err, "Versions", za0001)
 			return
-		}
-		switch msgp.UnsafeString(field) {
-		case "Volume":
-			z.Volume, err = dc.ReadString()
-			if err != nil {
-				err = msgp.WrapError(err, "Volume")
-				return
-			}
-		case "Name":
-			z.Name, err = dc.ReadString()
-			if err != nil {
-				err = msgp.WrapError(err, "Name")
-				return
-			}
-		case "IsEmptyDir":
-			z.IsEmptyDir, err = dc.ReadBool()
-			if err != nil {
-				err = msgp.WrapError(err, "IsEmptyDir")
-				return
-			}
-		case "LatestModTime":
-			z.LatestModTime, err = dc.ReadTime()
-			if err != nil {
-				err = msgp.WrapError(err, "LatestModTime")
-				return
-			}
-		case "Versions":
-			var zb0002 uint32
-			zb0002, err = dc.ReadArrayHeader()
-			if err != nil {
-				err = msgp.WrapError(err, "Versions")
-				return
-			}
-			if cap(z.Versions) >= int(zb0002) {
-				z.Versions = (z.Versions)[:zb0002]
-			} else {
-				z.Versions = make([]FileInfo, zb0002)
-			}
-			for za0001 := range z.Versions {
-				err = z.Versions[za0001].DecodeMsg(dc)
-				if err != nil {
-					err = msgp.WrapError(err, "Versions", za0001)
-					return
-				}
-			}
-		default:
-			err = dc.Skip()
-			if err != nil {
-				err = msgp.WrapError(err)
-				return
-			}
 		}
 	}
 	return
@@ -1178,9 +1170,8 @@ func (z *FileInfoVersions) DecodeMsg(dc *msgp.Reader) (err error) {
 
 // EncodeMsg implements msgp.Encodable
 func (z *FileInfoVersions) EncodeMsg(en *msgp.Writer) (err error) {
-	// map header, size 5
-	// write "Volume"
-	err = en.Append(0x85, 0xa6, 0x56, 0x6f, 0x6c, 0x75, 0x6d, 0x65)
+	// array header, size 4
+	err = en.Append(0x94)
 	if err != nil {
 		return
 	}
@@ -1189,39 +1180,14 @@ func (z *FileInfoVersions) EncodeMsg(en *msgp.Writer) (err error) {
 		err = msgp.WrapError(err, "Volume")
 		return
 	}
-	// write "Name"
-	err = en.Append(0xa4, 0x4e, 0x61, 0x6d, 0x65)
-	if err != nil {
-		return
-	}
 	err = en.WriteString(z.Name)
 	if err != nil {
 		err = msgp.WrapError(err, "Name")
 		return
 	}
-	// write "IsEmptyDir"
-	err = en.Append(0xaa, 0x49, 0x73, 0x45, 0x6d, 0x70, 0x74, 0x79, 0x44, 0x69, 0x72)
-	if err != nil {
-		return
-	}
-	err = en.WriteBool(z.IsEmptyDir)
-	if err != nil {
-		err = msgp.WrapError(err, "IsEmptyDir")
-		return
-	}
-	// write "LatestModTime"
-	err = en.Append(0xad, 0x4c, 0x61, 0x74, 0x65, 0x73, 0x74, 0x4d, 0x6f, 0x64, 0x54, 0x69, 0x6d, 0x65)
-	if err != nil {
-		return
-	}
 	err = en.WriteTime(z.LatestModTime)
 	if err != nil {
 		err = msgp.WrapError(err, "LatestModTime")
-		return
-	}
-	// write "Versions"
-	err = en.Append(0xa8, 0x56, 0x65, 0x72, 0x73, 0x69, 0x6f, 0x6e, 0x73)
-	if err != nil {
 		return
 	}
 	err = en.WriteArrayHeader(uint32(len(z.Versions)))
@@ -1242,21 +1208,11 @@ func (z *FileInfoVersions) EncodeMsg(en *msgp.Writer) (err error) {
 // MarshalMsg implements msgp.Marshaler
 func (z *FileInfoVersions) MarshalMsg(b []byte) (o []byte, err error) {
 	o = msgp.Require(b, z.Msgsize())
-	// map header, size 5
-	// string "Volume"
-	o = append(o, 0x85, 0xa6, 0x56, 0x6f, 0x6c, 0x75, 0x6d, 0x65)
+	// array header, size 4
+	o = append(o, 0x94)
 	o = msgp.AppendString(o, z.Volume)
-	// string "Name"
-	o = append(o, 0xa4, 0x4e, 0x61, 0x6d, 0x65)
 	o = msgp.AppendString(o, z.Name)
-	// string "IsEmptyDir"
-	o = append(o, 0xaa, 0x49, 0x73, 0x45, 0x6d, 0x70, 0x74, 0x79, 0x44, 0x69, 0x72)
-	o = msgp.AppendBool(o, z.IsEmptyDir)
-	// string "LatestModTime"
-	o = append(o, 0xad, 0x4c, 0x61, 0x74, 0x65, 0x73, 0x74, 0x4d, 0x6f, 0x64, 0x54, 0x69, 0x6d, 0x65)
 	o = msgp.AppendTime(o, z.LatestModTime)
-	// string "Versions"
-	o = append(o, 0xa8, 0x56, 0x65, 0x72, 0x73, 0x69, 0x6f, 0x6e, 0x73)
 	o = msgp.AppendArrayHeader(o, uint32(len(z.Versions)))
 	for za0001 := range z.Versions {
 		o, err = z.Versions[za0001].MarshalMsg(o)
@@ -1270,71 +1226,47 @@ func (z *FileInfoVersions) MarshalMsg(b []byte) (o []byte, err error) {
 
 // UnmarshalMsg implements msgp.Unmarshaler
 func (z *FileInfoVersions) UnmarshalMsg(bts []byte) (o []byte, err error) {
-	var field []byte
-	_ = field
 	var zb0001 uint32
-	zb0001, bts, err = msgp.ReadMapHeaderBytes(bts)
+	zb0001, bts, err = msgp.ReadArrayHeaderBytes(bts)
 	if err != nil {
 		err = msgp.WrapError(err)
 		return
 	}
-	for zb0001 > 0 {
-		zb0001--
-		field, bts, err = msgp.ReadMapKeyZC(bts)
+	if zb0001 != 4 {
+		err = msgp.ArrayError{Wanted: 4, Got: zb0001}
+		return
+	}
+	z.Volume, bts, err = msgp.ReadStringBytes(bts)
+	if err != nil {
+		err = msgp.WrapError(err, "Volume")
+		return
+	}
+	z.Name, bts, err = msgp.ReadStringBytes(bts)
+	if err != nil {
+		err = msgp.WrapError(err, "Name")
+		return
+	}
+	z.LatestModTime, bts, err = msgp.ReadTimeBytes(bts)
+	if err != nil {
+		err = msgp.WrapError(err, "LatestModTime")
+		return
+	}
+	var zb0002 uint32
+	zb0002, bts, err = msgp.ReadArrayHeaderBytes(bts)
+	if err != nil {
+		err = msgp.WrapError(err, "Versions")
+		return
+	}
+	if cap(z.Versions) >= int(zb0002) {
+		z.Versions = (z.Versions)[:zb0002]
+	} else {
+		z.Versions = make([]FileInfo, zb0002)
+	}
+	for za0001 := range z.Versions {
+		bts, err = z.Versions[za0001].UnmarshalMsg(bts)
 		if err != nil {
-			err = msgp.WrapError(err)
+			err = msgp.WrapError(err, "Versions", za0001)
 			return
-		}
-		switch msgp.UnsafeString(field) {
-		case "Volume":
-			z.Volume, bts, err = msgp.ReadStringBytes(bts)
-			if err != nil {
-				err = msgp.WrapError(err, "Volume")
-				return
-			}
-		case "Name":
-			z.Name, bts, err = msgp.ReadStringBytes(bts)
-			if err != nil {
-				err = msgp.WrapError(err, "Name")
-				return
-			}
-		case "IsEmptyDir":
-			z.IsEmptyDir, bts, err = msgp.ReadBoolBytes(bts)
-			if err != nil {
-				err = msgp.WrapError(err, "IsEmptyDir")
-				return
-			}
-		case "LatestModTime":
-			z.LatestModTime, bts, err = msgp.ReadTimeBytes(bts)
-			if err != nil {
-				err = msgp.WrapError(err, "LatestModTime")
-				return
-			}
-		case "Versions":
-			var zb0002 uint32
-			zb0002, bts, err = msgp.ReadArrayHeaderBytes(bts)
-			if err != nil {
-				err = msgp.WrapError(err, "Versions")
-				return
-			}
-			if cap(z.Versions) >= int(zb0002) {
-				z.Versions = (z.Versions)[:zb0002]
-			} else {
-				z.Versions = make([]FileInfo, zb0002)
-			}
-			for za0001 := range z.Versions {
-				bts, err = z.Versions[za0001].UnmarshalMsg(bts)
-				if err != nil {
-					err = msgp.WrapError(err, "Versions", za0001)
-					return
-				}
-			}
-		default:
-			bts, err = msgp.Skip(bts)
-			if err != nil {
-				err = msgp.WrapError(err)
-				return
-			}
 		}
 	}
 	o = bts
@@ -1343,7 +1275,7 @@ func (z *FileInfoVersions) UnmarshalMsg(bts []byte) (o []byte, err error) {
 
 // Msgsize returns an upper bound estimate of the number of bytes occupied by the serialized message
 func (z *FileInfoVersions) Msgsize() (s int) {
-	s = 1 + 7 + msgp.StringPrefixSize + len(z.Volume) + 5 + msgp.StringPrefixSize + len(z.Name) + 11 + msgp.BoolSize + 14 + msgp.TimeSize + 9 + msgp.ArrayHeaderSize
+	s = 1 + msgp.StringPrefixSize + len(z.Volume) + msgp.StringPrefixSize + len(z.Name) + msgp.TimeSize + msgp.ArrayHeaderSize
 	for za0001 := range z.Versions {
 		s += z.Versions[za0001].Msgsize()
 	}

--- a/cmd/storage-interface.go
+++ b/cmd/storage-interface.go
@@ -57,7 +57,7 @@ type StorageAPI interface {
 
 	// Metadata operations
 	DeleteVersion(ctx context.Context, volume, path string, fi FileInfo, forceDelMarker bool) error
-	DeleteVersions(ctx context.Context, volume string, versions []FileInfo) []error
+	DeleteVersions(ctx context.Context, volume string, versions []FileInfoVersions) []error
 	WriteMetadata(ctx context.Context, volume, path string, fi FileInfo) error
 	UpdateMetadata(ctx context.Context, volume, path string, fi FileInfo) error
 	ReadVersion(ctx context.Context, volume, path, versionID string, readData bool) (FileInfo, error)

--- a/cmd/storage-rest-client.go
+++ b/cmd/storage-rest-client.go
@@ -590,7 +590,7 @@ func (client *storageRESTClient) Delete(ctx context.Context, volume string, path
 }
 
 // DeleteVersions - deletes list of specified versions if present
-func (client *storageRESTClient) DeleteVersions(ctx context.Context, volume string, versions []FileInfo) (errs []error) {
+func (client *storageRESTClient) DeleteVersions(ctx context.Context, volume string, versions []FileInfoVersions) (errs []error) {
 	if len(versions) == 0 {
 		return errs
 	}

--- a/cmd/storage-rest-common.go
+++ b/cmd/storage-rest-common.go
@@ -18,7 +18,7 @@
 package cmd
 
 const (
-	storageRESTVersion       = "v40" // Add ReplicationState field
+	storageRESTVersion       = "v41" // Optimized DeleteVersions API
 	storageRESTVersionPrefix = SlashSeparator + storageRESTVersion
 	storageRESTPrefix        = minioReservedBucketPath + "/storage"
 )

--- a/cmd/storage-rest-server.go
+++ b/cmd/storage-rest-server.go
@@ -643,7 +643,7 @@ func (s *storageRESTServer) DeleteVersionsHandler(w http.ResponseWriter, r *http
 		return
 	}
 
-	versions := make([]FileInfo, totalVersions)
+	versions := make([]FileInfoVersions, totalVersions)
 	decoder := msgp.NewReader(r.Body)
 	for i := 0; i < totalVersions; i++ {
 		dst := &versions[i]

--- a/cmd/xl-storage-disk-id-check.go
+++ b/cmd/xl-storage-disk-id-check.go
@@ -421,8 +421,8 @@ func (p *xlStorageDiskIDCheck) Delete(ctx context.Context, volume string, path s
 
 // DeleteVersions deletes slice of versions, it can be same object
 // or multiple objects.
-func (p *xlStorageDiskIDCheck) DeleteVersions(ctx context.Context, volume string, versions []FileInfo) (errs []error) {
-	// Mererly for tracing storage
+func (p *xlStorageDiskIDCheck) DeleteVersions(ctx context.Context, volume string, versions []FileInfoVersions) (errs []error) {
+	// Merely for tracing storage
 	path := ""
 	if len(versions) > 0 {
 		path = versions[0].Name

--- a/cmd/xl-storage.go
+++ b/cmd/xl-storage.go
@@ -821,13 +821,102 @@ func (s *xlStorage) ListDir(ctx context.Context, volume, dirPath string, count i
 	return entries, nil
 }
 
+func (s *xlStorage) deleteVersions(ctx context.Context, volume, path string, fis ...FileInfo) error {
+	buf, err := s.ReadAll(ctx, volume, pathJoin(path, xlStorageFormatFile))
+	if err != nil {
+		if err != errFileNotFound {
+			return err
+		}
+		metaDataPoolPut(buf) // Never used, return it
+
+		buf, err = s.ReadAll(ctx, volume, pathJoin(path, xlStorageFormatFileV1))
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(buf) == 0 {
+		return errFileNotFound
+	}
+
+	volumeDir, err := s.getVolDir(volume)
+	if err != nil {
+		return err
+	}
+
+	if !isXL2V1Format(buf) {
+		// Delete the meta file, if there are no more versions the
+		// top level parent is automatically removed.
+		return s.deleteFile(volumeDir, pathJoin(volumeDir, path), true)
+	}
+
+	var xlMeta xlMetaV2
+	if err = xlMeta.Load(buf); err != nil {
+		return err
+	}
+
+	var (
+		dataDir     string
+		lastVersion bool
+	)
+
+	for _, fi := range fis {
+		dataDir, lastVersion, err = xlMeta.DeleteVersion(fi)
+		if err != nil {
+			return err
+		}
+		if dataDir != "" {
+			versionID := fi.VersionID
+			if versionID == "" {
+				versionID = nullVersionID
+			}
+			// PR #11758 used DataDir, preserve it
+			// for users who might have used master
+			// branch
+			if !xlMeta.data.remove(versionID, dataDir) {
+				filePath := pathJoin(volumeDir, path, dataDir)
+				if err = checkPathLength(filePath); err != nil {
+					return err
+				}
+				if err = s.moveToTrash(filePath, true); err != nil {
+					if err != errFileNotFound {
+						return err
+					}
+				}
+			}
+		}
+	}
+
+	if !lastVersion {
+		buf, err = xlMeta.AppendTo(metaDataPoolGet())
+		defer metaDataPoolPut(buf)
+		if err != nil {
+			return err
+		}
+
+		return s.WriteAll(ctx, volume, pathJoin(path, xlStorageFormatFile), buf)
+	}
+
+	// Move xl.meta to trash
+	filePath := pathJoin(volumeDir, path, xlStorageFormatFile)
+	if err = checkPathLength(filePath); err != nil {
+		return err
+	}
+
+	err = s.moveToTrash(filePath, false)
+	if err == nil || err == errFileNotFound {
+		s.deleteFile(volumeDir, pathJoin(volumeDir, path), false)
+	}
+	return err
+}
+
 // DeleteVersions deletes slice of versions, it can be same object
 // or multiple objects.
-func (s *xlStorage) DeleteVersions(ctx context.Context, volume string, versions []FileInfo) []error {
+func (s *xlStorage) DeleteVersions(ctx context.Context, volume string, versions []FileInfoVersions) []error {
 	errs := make([]error, len(versions))
 
-	for i, version := range versions {
-		if err := s.DeleteVersion(ctx, volume, version.Name, version, false); err != nil {
+	for i, fiv := range versions {
+		if err := s.deleteVersions(ctx, volume, fiv.Name, fiv.Versions...); err != nil {
 			errs[i] = err
 		}
 	}

--- a/go.mod
+++ b/go.mod
@@ -63,7 +63,7 @@ require (
 	github.com/nats-io/stan.go v0.8.3
 	github.com/ncw/directio v1.0.5
 	github.com/nsqio/go-nsq v1.0.8
-	github.com/philhofer/fwd v1.1.1
+	github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
 	github.com/pierrec/lz4 v2.6.0+incompatible
 	github.com/pkg/errors v0.9.1
 	github.com/prometheus/client_golang v1.11.0
@@ -74,7 +74,7 @@ require (
 	github.com/secure-io/sio-go v0.3.1
 	github.com/shirou/gopsutil/v3 v3.21.9
 	github.com/streadway/amqp v1.0.0
-	github.com/tinylib/msgp v1.1.6
+	github.com/tinylib/msgp v1.1.7-0.20211026165309-e818a1881b0e
 	github.com/valyala/bytebufferpool v1.0.0
 	github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
 	github.com/yargevad/filepathx v1.0.0

